File: 2023_8/baekjoon/25682.py
import sys

N, M, K = map(int, sys.stdin.readline().split())
arr = [list(str(sys.stdin.readline().strip())) for _ in range(N)]


def solution(color):
    pre_sum = [[0 for _ in range(M + 1)] for _ in range(N + 1)]
    for i in range(N):
        for j in range(M):
            if (i + j) % 2 == 0:
                if arr[i][j] != color:
                    value = 1
                else:
                    value = 0
            else:
                if arr[i][j] == color:
                    value = 1
                else:
                    value = 0
            pre_sum[i + 1][j + 1] = pre_sum[i][j + 1] + \
                pre_sum[i + 1][j] - pre_sum[i][j] + value
    result = N * M
    for i in range(1, N - K + 2):
        for j in range(1, M - K + 2):
            result = min(result, pre_sum[i + K - 1][j + K - 1] - pre_sum[i - 1]
                         [j + K - 1] - pre_sum[i + K - 1][j - 1] + pre_sum[i - 1][j - 1])
    return result


print(min(solution('B'), solution('W')))

File: 2023_8/baekjoon/25682.rs
use std::io::{stdin, BufRead, Read};

struct Info {
    n: usize,
    m: usize,
    k: usize,
}

fn main() {
    let mut buf = String::new();
    stdin().lock().read_line(&mut buf).unwrap();
    let infos: Info = {
        let parts: Vec<usize> = buf
            .split_ascii_whitespace()
            .map(|s| s.parse::<usize>().unwrap())
            .collect();
        Info {
            n: parts[0],
            m: parts[1],
            k: parts[2],
        }
    };
    buf.clear();
    stdin().read_to_string(&mut buf).unwrap();
    let arr: Vec<Vec<char>> = buf
        .lines()
        .skip(0)
        .map(|line| line.chars().collect())
        .collect();

    println!(
        "{}",
        solution(&infos, 'W', &arr).min(solution(&infos, 'B', &arr))
    );
}

fn solution(info: &Info, color: char, arr: &Vec<Vec<char>>) -> usize {
    let mut dp = vec![vec![0; info.m + 1]; info.n + 1];
    let mut result = info.n * info.m;

    for i in 0..info.n {
        for j in 0..info.m {
            let value;
            if (i + j) % 2 == 0 {
                value = (color != arr[i][j]) as usize;
            } else {
                value = (color == arr[i][j]) as usize;
            }

            dp[i + 1][j + 1] = dp[i + 1][j] + dp[i][j + 1] - dp[i][j] + value;
        }
    }

    for i in 1..(info.n - info.k + 2) {
        for j in 1..(info.m - info.k + 2) {
            result = result.min(
                dp[i + info.k - 1][j + info.k - 1]
                    - dp[i - 1][j + info.k - 1]
                    - dp[i + info.k - 1][j - 1]
                    + dp[i - 1][j - 1],
            )
        }
    }
    result
}
